connect-inject/handler.go

```diff
@@ -393,6 +478,76 @@ func (h *Handler) defaultAnnotations(pod *corev1.Pod, patches *[]jsonpatch.JsonP
 	return nil
 }
 
+func (h *Handler) consulNamespace(ns string) string {
+	// Return an empty string if namespaces aren't enabled. This
+	// functionality is used for init and sidecar container creation.
+	if !h.EnableNamespaces {
+		return ""
+	}
+
+	// Mirroring takes precedence
+	if h.EnableNSMirroring {
+		return fmt.Sprintf("%s%s", h.MirroringPrefix, ns)
+	} else {
+		return h.ConsulNSName
+	}
+}
+
+func (h *Handler) checkAndCreateNamespace(ns string) error {
+	// Check if the Consul namespace exists
+	namespaceInfo, _, err := h.ConsulClient.Namespaces().Read(ns, nil)
+	if err != nil {
+		return err
+	}
+
+	// If not, create it
+	if namespaceInfo == nil {
+		consulNamespace := api.Namespace{
+			Name:        ns,
+			Description: "Auto-generated by a Connect Injector",
+			// TODO: when metadata is added to the api
+			// Meta:        map[string]string{"external-source": "kubernetes"},
+		}
+
+		_, _, err = h.ConsulClient.Namespaces().Create(&consulNamespace, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *Handler) checkAndCreateBindingRule(ns string) error {
+	// Check if the binding rule exists in this namespace
+	existingRules, _, err := h.ConsulClient.ACL().BindingRuleList(h.AuthMethod, &api.QueryOptions{Namespace: ns})
+	if err != nil {
+		return err
+	}
+
+	// When there are no binding rules for the current namespace, one needs to be created
+	if len(existingRules) == 0 {
+		// Note: This description uses the AuthMethod name which differs from binding rules
+		// create in the ACL bootstrap method where the release name is used. It didn't seem
+		// important enough to add yet another flag to make the distinction, given that the
+		// AuthMethod name will include that info if using the ACL bootstrapping process.
+		abr := api.ACLBindingRule{
+			Description: fmt.Sprintf("Consul %s default binding rule", h.AuthMethod),
+			AuthMethod:  h.AuthMethod,
+			BindType:    api.BindingRuleBindTypeService,
+			BindName:    "${serviceaccount.name}",
+			Selector:    h.BindingRuleSelector,
+			Namespace:   ns,
+		}
+		_, _, err = h.ConsulClient.ACL().BindingRuleCreate(&abr, nil)
+		if err != nil {
```

Comment:
could just return `err` here ;)

Reply:
Definitely could! I'm going to leave it because I like the fall through `return nil` to apply to the whole function.